### PR TITLE
ARROW-1636: [C++][Integration] Implement integration test parsing in C++ for null type, add integration test data generation

### PR DESCRIPTION
**Pull request by wesm**
This enables integration testing of null type for C++ and Java. I was a bit surprised that the cross-compatibility tests passed so I would appreciate a sanity check from a Java developer to see if there are integration tests covering the integration-JSON-format parsing and null type IPC round trip. If there are not, let's ensure there are relevant JIRA issues so it can be tested and kept working

### CODE REVIEW

**Comment by tianchen92** [comment]
> This enables integration testing of null type for C++ and Java. I was a bit surprised that the cross-compatibility tests passed so I would appreciate a sanity check from a Java developer to see if there are integration tests covering the integration-JSON-format parsing and null type IPC round trip. If there are not, let's ensure there are relevant JIRA issues so it can be tested and kept working

Hi, Wes, I think I had add JSON test for null type in this PR https://github.com/apache/arrow/pull/5164/files

File: dev/archery/archery/integration/datagen.py [hunk 1]
```diff
@@ -1075,12 +1095,27 @@ def _temp_path():
...
+        generate_decimal_case().skip_category('Go'),
+
         generate_datetime_case(),
-        generate_interval_case(),
-        generate_map_case(),
+
+        # TODO(ARROW-5239): Intervals + JS
+        generate_interval_case().skip_category('JS'),
+
+        # TODO(ARROW-3679): Map + Go
```
Comment (nealrichardson) [inline dev/archery/archery/integration/datagen.py]:
```suggestion
        # TODO(ARROW-5620): Map + Go
```

File: dev/archery/archery/integration/datagen.py [hunk 2]
```diff
@@ -1075,12 +1095,27 @@ def _temp_path():
...
         generate_primitive_case([17, 20], name='primitive'),
         generate_primitive_case([0, 0, 0], name='primitive_zerolength'),
-        generate_decimal_case(),
+
+        generate_null_case([10, 0])
+        .skip_category('JS')
+        .skip_category('Go'),
+
+        # TODO(ARROW-3676): Go tests
+        generate_decimal_case().skip_category('Go'),
```
Comment (nealrichardson) [inline dev/archery/archery/integration/datagen.py]: ARROW-3676 is resolved

```suggestion
        generate_decimal_case(),
```

File: dev/archery/archery/integration/datagen.py [hunk 3]
```diff
@@ -1075,12 +1095,27 @@ def _temp_path():
         generate_primitive_case([], name='primitive_no_batches'),
         generate_primitive_case([17, 20], name='primitive'),
         generate_primitive_case([0, 0, 0], name='primitive_zerolength'),
-        generate_decimal_case(),
+
+        generate_null_case([10, 0])
+        .skip_category('JS')
```
Comment (nealrichardson) [inline dev/archery/archery/integration/datagen.py]: We should make Jiras for these skips. Also it appears that Java is not implemented (the test fails) so we should skip Java too.

**Review by nealrichardson — COMMENTED** [review]
A couple of notes

**Comment by nealrichardson** [comment]
The null type C++ to Java test failed: https://github.com/apache/arrow/pull/6368/checks?check_run_id=428756741#step:5:4858

**Comment by wesm** [comment]
@lidavidm @liyafan82 @tianchen92 @emkornfield the "generated_null" case fails in the Flight integration tests but passes in the IPC test (i.e. when you omit `--run-flight`). The failure is

```
With output:
--------------
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by io.netty.util.internal.ReflectionUtil (file:/home/wesm/code/arrow/java/flight/flight-core/target/flight-core-1.0.0-SNAPSHOT-jar-with-dependencies.jar) to constructor java.nio.DirectByteBuffer(long,int)
WARNING: Please consider reporting this to the maintainers of io.netty.util.internal.ReflectionUtil
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
Verifying location grpc+tcp://localhost:35839
Exception in thread "main" java.lang.RuntimeException: java.lang.IllegalArgumentException: Different row count:
0 != 10
	at org.apache.arrow.flight.example.integration.IntegrationTestClient.testStream(IntegrationTestClient.java:175)
	at org.apache.arrow.flight.example.integration.IntegrationTestClient.run(IntegrationTestClient.java:95)
	at org.apache.arrow.flight.example.integration.IntegrationTestClient.main(IntegrationTestClient.java:68)
Caused by: java.lang.IllegalArgumentException: Different row count:
0 != 10
	at org.apache.arrow.vector.util.Validator.compareVectorSchemaRoot(Validator.java:98)
	at org.apache.arrow.flight.example.integration.IntegrationTestClient.testStream(IntegrationTestClient.java:173)
	... 2 more
```

This implies there is a missing assertion on the IPC side of the test. Can someone help?

**Comment by lidavidm** [comment]
https://github.com/apache/arrow/blob/master/java/tools/src/main/java/org/apache/arrow/tools/Integration.java#L208

If I'm not mistaken, this means that we only compare the data fully if there's actual data in both JSON and in the Arrow file?

Though the Flight test is also potentially wrong: 

https://github.com/apache/arrow/blob/master/java/flight/flight-core/src/main/java/org/apache/arrow/flight/example/integration/IntegrationTestClient.java#L166-L173

It only compares the last batch sent over the wire.

**Comment by liyafan82** [comment]
According to @lidavidm's investigation, maybe we need another issue to resolve the defects in the tests?

File: dev/archery/archery/integration/datagen.py [hunk 4]
```diff
@@ -1075,12 +1095,27 @@ def _temp_path():
...
+        # TODO(ARROW-5239): Intervals + JS
+        generate_interval_case().skip_category('JS'),
+
+        # TODO(ARROW-3679): Map + Go
+        generate_map_case().skip_category('Go'),
+
         generate_nested_case(),
-        generate_dictionary_case(),
+
+        # TODO(ARROW-3039): Dictionaries in GO
```
Comment (nealrichardson) [inline dev/archery/archery/integration/datagen.py]:
```suggestion
        # TODO(ARROW-3039, ARROW-5267): Dictionaries in GO
```

File: dev/archery/archery/integration/datagen.py (hunk 3 above)
Comment (nealrichardson) [reply dev/archery/archery/integration/datagen.py]:
```suggestion
        .skip_category('Java') # TODO(ARROW-7899)
        .skip_category('JS') # TODO(ARROW-7900)
```

File: dev/archery/archery/integration/datagen.py [hunk 5]
```diff
@@ -1075,12 +1095,27 @@ def _temp_path():
         generate_primitive_case([], name='primitive_no_batches'),
         generate_primitive_case([17, 20], name='primitive'),
         generate_primitive_case([0, 0, 0], name='primitive_zerolength'),
-        generate_decimal_case(),
+
+        generate_null_case([10, 0])
+        .skip_category('JS')
+        .skip_category('Go'),
```
Comment (nealrichardson) [inline dev/archery/archery/integration/datagen.py]:
```suggestion
        .skip_category('Go'), # TODO(ARROW-7901)
```

File: dev/archery/archery/integration/datagen.py [hunk 6]
```diff
@@ -1075,12 +1095,27 @@ def _temp_path():
...
+
+        # TODO(ARROW-3679): Map + Go
+        generate_map_case().skip_category('Go'),
+
         generate_nested_case(),
-        generate_dictionary_case(),
+
+        # TODO(ARROW-3039): Dictionaries in GO
+        generate_dictionary_case().skip_category('Go'),
+
```
Comment (nealrichardson) [inline dev/archery/archery/integration/datagen.py]:
```suggestion

        # TODO(ARROW-7902)
```

**Review by nealrichardson — COMMENTED** [review]
I created some Jiras, including for fixing the Java null test, and added TODOs for them in PR suggestions

**Comment by nealrichardson** [comment]
I've made some more Jiras and linked them on https://docs.google.com/spreadsheets/d/1Yu68rn2XMBpAArUfCOP9LC7uHb06CQrtqKE5vQ4bQx4/edit#gid=0

**Comment by BryanCutler** [comment]
>I created some Jiras, including for fixing the Java null test, and added TODOs for them in PR suggestions

I can start looking into the Java issues tomorrow

**Comment by BryanCutler** [comment]
I merged in a fix for Java, so integration tests should pass here now

**Comment by pitrou** [comment]
Rebased.

File: dev/archery/archery/integration/datagen.py (hunk 3 above)
Comment (pitrou) [reply dev/archery/archery/integration/datagen.py]: ARROW-7899 is fixed now.

File: dev/archery/archery/integration/datagen.py (hunk 2 above)
Comment (pitrou) [reply dev/archery/archery/integration/datagen.py]: Unfortunately, tests fail: [ARROW-7948](https://issues.apache.org/jira/browse/ARROW-7948)

**Review by pitrou — APPROVED** [review]
+1, will merge if green

**Comment by wesm** [comment]
Thanks all

**Pull request closed without merging**